pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/loader/SchemaChangeHandler.java

```diff
@@ -0,0 +1,341 @@
+    newMetadataProperties.save(newMetadataFile);
+
+    // delete the temporary directory
+    FileUtils.deleteDirectory(tmpIndexDir);
+    return new SegmentMetadataImpl(indexDir);
+  }
+
+  /**
+   * Computes the action needed for each column.
+   * This method compares the column metadata across schema and segment
+   * @return Action Map for each column
+   */
+  private Map<String, Action> computeSchemaUpdateAction() {
+    Collection<String> columnsInSchema = schema.getColumnNames();
+    Map<String, Action> schemaUpdateActionMap = new LinkedHashMap<>();
+    for (String column : columnsInSchema) {
+      FieldSpec fieldSpec = schema.getFieldSpecFor(column);
+      ColumnMetadata columnMetadata = metadata.getColumnMetadataFor(column);
+      if (columnMetadata != null) {
+        boolean isAutoGenerated = columnMetadata.isAutoGenerated();
+        String defaultValueFromMetadata = String.valueOf(columnMetadata.getDefaultNullValue());
+        String defaultValueFromSchema = String.valueOf(fieldSpec.getDefaultNullValue());
+        if (isAutoGenerated && !defaultValueFromMetadata.equals(defaultValueFromSchema)) {
+          if (fieldSpec.getFieldType() == FieldType.DIMENSION) {
+            schemaUpdateActionMap.put(column, Action.UPDATE_DIMENSION_DEFAULT_VALUE);
+          } else if (fieldSpec.getFieldType() == FieldType.METRIC) {
+            schemaUpdateActionMap.put(column, Action.UPDATE_METRIC_DEFAULT_VALUE);
+          } else {
+            throw new UnsupportedOperationException(
+                "Adding time column is not supported in schema evolution");
+          }
+        }
+      } else {
+        if (fieldSpec.getFieldType() == FieldType.DIMENSION) {
+          schemaUpdateActionMap.put(column, Action.ADD_DIMENSION);
+        } else if (fieldSpec.getFieldType() == FieldType.METRIC) {
+          schemaUpdateActionMap.put(column, Action.ADD_METRIC);
+        } else {
+          throw new UnsupportedOperationException(
+              "Adding time column is not supported in schema evolution");
+        }
+      }
+    }
+    return schemaUpdateActionMap;
+  }
+
+  @SuppressWarnings("unchecked")
+  private void createColumnWithDefaultValue(String column,
+      PropertiesConfiguration metadataProperties, File tmpIndexDir) throws Exception {
+    int totalDocs = metadata.getTotalDocs();
+    int totalRawDocs = metadata.getTotalRawDocs();
+    int totalAggDocs = metadata.getTotalDocs() - metadata.getTotalRawDocs();
+    // we need only 1 bit since we have only 1 value in the forward index
+    FieldSpec fieldSpec = schema.getFieldSpecFor(column);
+    int dictionaryElementSize;
+    Object min = fieldSpec.getDefaultNullValue();
+    Object max = fieldSpec.getDefaultNullValue();
+    Object sortedArray;
+    switch (fieldSpec.getDataType()) {
+    case STRING:
+    case BOOLEAN:
+      byte[] bytes = fieldSpec.getDefaultNullValue().toString().getBytes();
```

Comment:
That's the discussion between JF and me. If no default null value explicitly set by the user, when calling getDefaultNullValue() on boolean type field spec will through exceptions. The reason for this is that there is no proper default null value for boolean type. For this use case, user should already set the default null value for boolean explicitly.